test/codec-test.hs

```diff
@@ -0,0 +1,271 @@
+{-# LANGUAGE CPP                        #-}
+{-# LANGUAGE DerivingStrategies         #-}
+{-# LANGUAGE GeneralisedNewtypeDeriving #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RankNTypes                 #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# OPTIONS_GHC -fspecialize-aggressively #-}
+module Main (main) where
+
+import           Control.Monad (when)
+import           Control.Monad.ST.Strict (ST, runST)
+import           Data.Foldable (fold)
+import           Data.STRef
+import           Data.Vector (Vector)
+import qualified Data.Vector as V
+import           Database.LSMTree.Common (BloomFilterAlloc (..),
+                     DiskCachePolicy (..), NumEntries (..), TableConfig (..),
+                     WriteBufferAlloc (..), defaultTableConfig)
+import           Database.LSMTree.Internal.MergeSchedule (NumRuns (..))
+import           Database.LSMTree.Internal.RunNumber (RunNumber (..))
+import           Database.LSMTree.Internal.Snapshot
+import           Database.LSMTree.Internal.Snapshot.Codec
+import qualified System.FS.API as FS
+import           System.FS.API.Types (FsPath, MountPoint (..), fsToFilePath,
+                     mkFsPath, (<.>))
+import           System.FS.IO (HandleIO, ioHasFS)
+import qualified Test.Tasty as Tasty
+import           Test.Tasty (TestName, TestTree, testGroup)
+import qualified Test.Tasty.Golden as Au
+import           Text.Printf (printf)
+
+
+-- |
+-- Delete output files on test-case success.
+-- Change the option here if this is undesireable.
+handleOutputFiles :: TestTree -> TestTree
+handleOutputFiles = Tasty.localOption Au.OnPass
+
+-- |
+--
+-- Internally, the function will infer the correct filepath names.
+--
+-- There are three paths for both the checksum and the snapshot files:
+--   1. The filepath of type @FsPath@ to which data is written.
+--   2. The filepath of type @FilePath@ from which data is read.
+--   3. The filepath of type @FilePath@ against which the data is compared.
+--
+-- These file types' bindings have the following infix annotations, respectively:
+--   1. (Fs) for FsPath
+--   2. (Hs) for "Haskell" path
+--   3. (Au) for "Golden file" path
```

Comment:
This should be a comment on the let binding inside the function, not on the function itself. It concerns itself with implementation details

Reply:
Moved to the appropriate code location.